src/io/fasta.rs

```diff
@@ -321,7 +316,7 @@ impl<R: io::Read + io::Seek> IndexedReader<R> {
     }
 
     /// Return the IndexRecord for the given sequence name or io::Result::Err
-    fn idx_by_index(&self, seq_index: usize) -> io::Result<IndexRecord> {
+    fn idx_by_rid(&self, seq_index: usize) -> io::Result<IndexRecord> {
         match self.index.seqs.get(seq_index) {
             Some(name) => self.idx(&name),
```

Comment:
Mhm, this looks a bit weird to me. We provide an index, then we get a name, and with the name we look up an index again? Why that?

Reply:
The main data structures in the `Index` struct used by `IndexedReader` are a HashMap that maps sequence names to idx objects and a Vector of sequence names.

    pub struct Index {
        inner: collections::HashMap<String, IndexRecord>,
        seqs: Vec<String>,
    }

I chose this to minimize the additional memory required and to minimize changes to the code. Alternatively we could add a `Vec<IndexRecord>` or `HashMap<usize, IndexRecord>` to directly map positions to index record. This would eliminate one layer of indirection at the expense of more memory.

However, this might be a naming issue. I renamed seq_index to `rid`.

Reply:
What about `inner: Vec<IndexRecord>` and `name_to_rid: HashMap<String, usize>`. This gives us constant time access for both RID and name.

Reply:
Done. This made the `sequences` method a little more complex (now sorting is required to restore the order of the records in the file).

Querying the Index with an RID is just indexing into the `inner` Vector, while querying with a name requires a HashMap lookup to get the rid and then index into the Vector.

Reply:
Can't you add the name to the index record struct, and let the vector be a vector of references to the index records from the hashmap?